File: GO/a2zSheet/step17_tries/count_distinct_substrings.cpp
struct Node{
    Node* links[26];
};

int countDistinctSubstrings(string &s)
{   int n = s.size();
    Node* root = new Node();
    int cnt = 0;
    for(int i=0;i<n;i++){
        Node* temp = root;
        for(int j = i;j<n;j++){
            if(!temp->links[s[j]-'a']){
                cnt++;
                temp->links[s[j]-'a'] = new Node();
            }
            temp = temp->links[s[j]-'a'];
        }
    }

    return cnt+1;
}

File: a2zSheet/step17_tries/count_distinct_substrings.go
package main

import "fmt"

type Node struct {
	links [26]*Node
}

func countDistinctSubstrings(s string) int {
	n := len(s)
	root := &Node{}
	cnt := 0
	for i := 0; i < n; i++ {
		temp := root
		for j := i; j < n; j++ {
			if temp.links[s[j]-'a'] == nil {
				cnt++
				temp.links[s[j]-'a'] = &Node{}
			}
			temp = temp.links[s[j]-'a']
		}
	}

	return cnt + 1
}

func main() {
	s := "abc"
	fmt.Print(countDistinctSubstrings(s))
}
